librz/bin/dwarf.c

```diff
@@ -449,133 +468,125 @@ static const ut8 *parse_line_header_source(RzBinFile *bf, const ut8 *buf, const
+		entry->id_idx = id_idx;
+		entry->mod_time = mod_time;
+		entry->file_len = file_len;
+	}
+	hdr->file_names_count = rz_vector_len(&file_names);
+	hdr->file_names = rz_vector_flush(&file_names);
+	rz_vector_fini(&file_names);
 
-			if (!len) {
-				buf++;
-				break;
-			}
-			buf += len + 1;
-			if (buf >= buf_end) {
-				buf = NULL;
-				goto beach;
-			}
-			buf = rz_uleb128(buf, buf_end - buf, &id_idx, NULL);
-			if (buf >= buf_end) {
-				buf = NULL;
-				goto beach;
-			}
-			buf = rz_uleb128(buf, buf_end - buf, &mod_time, NULL);
-			if (buf >= buf_end) {
-				buf = NULL;
-				goto beach;
-			}
-			buf = rz_uleb128(buf, buf_end - buf, &file_len, NULL);
-			if (buf >= buf_end) {
-				buf = NULL;
-				goto beach;
-			}
+beach:
+	return buf;
+}
 
-			if (i) {
-				char *include_dir = NULL, *comp_dir = NULL, *pinclude_dir = NULL;
-				if (id_idx > 0) {
-					include_dir = pinclude_dir = sdb_array_get(sdb, "includedirs", id_idx - 1, 0);
-					if (include_dir && include_dir[0] != '/') {
-						comp_dir = sdb_get(bf->sdb_addrinfo, "DW_AT_comp_dir", 0);
-						if (comp_dir) {
-							include_dir = rz_str_newf("%s/%s/", comp_dir, include_dir);
-						}
-					}
-				} else {
-					include_dir = pinclude_dir = sdb_get(bf->sdb_addrinfo, "DW_AT_comp_dir", 0);
-					if (!include_dir) {
-						include_dir = "./";
-					}
-				}
-
-				if (hdr->file_names) {
-					hdr->file_names[count].name = rz_str_newf("%s/%s", include_dir ? include_dir : "", filename);
-					hdr->file_names[count].id_idx = id_idx;
-					hdr->file_names[count].mod_time = mod_time;
-					hdr->file_names[count].file_len = file_len;
-				}
-				if (comp_dir) {
-					RZ_FREE(include_dir);
-					RZ_FREE(comp_dir);
-				}
-				RZ_FREE(pinclude_dir);
-			}
-			count++;
-			if (mode == RZ_MODE_PRINT && i) {
-				print("  %d     %" PFMT64d "       %" PFMT64d "         %" PFMT64d "          %s\n", entry_index++, id_idx, mod_time, file_len, filename);
-			}
-		}
-		if (i == 0) {
-			if (count > 0) {
-				hdr->file_names = calloc(sizeof(file_entry), count);
-			} else {
-				hdr->file_names = NULL;
+RZ_API char *rz_bin_dwarf_line_header_get_full_file_path(Sdb *sdb_addrinfo, const RzBinDwarfLineHeader *header, ut64 file_index) {
+	if (file_index >= header->file_names_count) {
+		return NULL;
+	}
+	RzBinDwarfLineFileEntry *file = &header->file_names[file_index];
+	rz_return_val_if_fail(header && file, NULL);
+	const char *include_dir = NULL, *comp_dir = NULL;
+	if (file->id_idx > 0 && file->id_idx - 1 < header->include_dirs_count) {
+		include_dir = header->include_dirs[file->id_idx - 1];
+		if (include_dir && include_dir[0] != '/') {
+			comp_dir = sdb_const_get(sdb_addrinfo, "DW_AT_comp_dir", 0);
+			if (comp_dir) {
+				include_dir = rz_str_newf("%s/%s/", comp_dir, include_dir);
 			}
-			hdr->file_names_count = count;
-			buf = tmp_buf;
-			count = 0;
 		}
+	} else {
+		include_dir = sdb_const_get(sdb_addrinfo, "DW_AT_comp_dir", 0);
+	}
+	if (!include_dir) {
+		include_dir = "./";
```

Comment:
what if you have windows?

Reply:
Dwarf standard does not seem to specify the exact format of paths so apparently it is indeed target-dependent. However the one Windows binary we have that also contains dwarf, which is `test/bins/pe/hello_world_not_stripped.exe`, contains path names like `/private/tmp/mingw-w64-20180808-82873-149eclj/mingw-w64-v5.0.4/mingw-w64-crt/build-i686/../crt//crtexe.c` with slashes so if we really encounter a file that has backslashes there it will probably be better to only use slashes in this place and later just treat slashes and backslashes equally in the result because we can't really detect that.

Reply:
We need more samples then for MinGW-compiled files with DWARF then.

Reply:
Please add a comment here then.